j5/backends/backend.py

```diff
@@ -161,16 +137,20 @@ def __init__(self, name: str):
 
     @property
     def supported_boards(self) -> Set[Type['Board']]:
-        """The boards that are supported by this backend group."""
+        """The boards that are supported by this environment."""
         return set(self.board_backend_mapping.keys())
 
     def __str__(self) -> str:
         """Get a string representation of this group."""
         return self.name
 
-    def register_backend(self, board: Type['Board'], backend: Type[Backend]) -> None:
-        """Register a new backend with this Backend Group."""
-        self.board_backend_mapping[board] = backend
+    def register_backend(self, backend: Type[Backend]) -> None:
+        """Register a new backend with this environment."""
+        board_type: Type['Board'] = cast(Type['Board'], backend.board)
```

Comment:
`j5/backends/backend.py:149: error: Incompatible types in assignment (expression has type "Callable[[Backend], Type[Board]]", variable has type "Type[Board]")`

This is caused by `mypy` treating a `property` as different to the concrete type.

```
    def register_backend(self, backend: Type[Backend]) -> None:
        """Register a new backend with this environment."""
        reveal_type(backend.board)
		...
```

Revealing the type (above) yields the following:

`j5/backends/backend.py:149: note: Revealed type is 'def (self: j5.backends.backend.Backend) -> Type[j5.boards.board.Board]'`

Reply:
I think the issue is partially due to `backend` being of type `Type[Backend]` rather than 
`Backend`. If the below code is used  MyPy is happy but it will almost certainly run time error when we don't provide the necessary values to the subtype's constructor.
```
board_type: Type['Board'] = backend().board
```
Converting `Backend.board` from a `@property` to a `@classmethod` is the only thing, other than the `cast`, that I can think of but that would require large changes to the codebase and `Backend.board()` doesn't feel particularly nice, though a rename could fix that.

Do you think it's worth changing, or should we just leave as the `cast`?

Reply:
I think it's worth changing, but the changes required are out of scope of this PR. #409 is tracking this.